diesel_derives/src/lib.rs

```diff
@@ -394,10 +394,31 @@ pub fn derive_query_id(input: TokenStream) -> TokenStream {
 ///
 /// This trait can only be derived for structs, not enums.
 ///
-/// **When this trait is derived, it will assume that the order of fields on your
-/// struct match the order of the fields in the query. This means that field
-/// order is significant if you are using `#[derive(Queryable)]`. Field name has
-/// no effect.**
+/// **When this trait is derived, it will assume that __all fields on your
+/// struct__ is the same as __all fields in the query__, including the order.
+/// This means that field order is significant if you are using
+/// `#[derive(Queryable)]`. Field name has no effect. For example,
+///
+/// ```rust
+/// # extern crate diesel;
+/// # use diesel::prelude::*;
+/// table! {
+///     posts {
+///         id -> Integer,
+///         user_id -> Integer,
+///         title -> VarChar,
+///     }
+/// }
+///
+/// #[derive(Queryable, Debug)]
+/// pub struct Post {
+///     id: i32,
+///     user_id: i32,
+///     title: String,
+///     // this cannot be id, user_id (missing title)
+///     // this cannot be id, title, user_id (incorrect order)
+/// }
+/// ```
 ///
```

Comment:
I think that the example should also show calculated fields, because for me it is still unclear why diesel can't at compile time calculate for each field in struct correct index of it in the `table!` macro. this should be as simple as get field name and find its index in the list of fields declared in the `table!` macro

Reply:
I didn't know how calculated fields works, maybe you can just provide a suggestion?

Reply:
@Mingun Please stop making random claims without having something to show. As you seem to be sure that this is just easy, it shouldn't be to hard for you to explain on the following example how this would work:

```rust
table! {
    users (id) {
        id -> Integer,
        name -> Text,
    }
}

table! {
    posts(id) {
        id -> Integer,
        title -> Text,
        body -> Nullable<Text>,
        user_id -> Integer,
    }
}

joinable!(…);
allow_tables_to_appear_in_same_query!(…);

#[derive(Queryable)]
struct MyResult {
    user_id: i32,
    post_id: Option<i32>,
    some_function_result: i32,
    some_integer_constant: i32
}

sql_function!(fn foo(a: Text, b: Integer) -> Integer);

fn some_not_so_complex_query(conn: &PgConnection) -> QueryResult<MyResult> {
    users::table.left_join(posts::table)
        .select((users::id, posts::id.nullable(), foo(users::name, users::id), 42.into_sql::<Integer>()))
        .load(conn)
}
```

Especially I want to know how you would address the following points:
* Handle that a column coming via a left join is nullable
* Handle the results of sql functions (or any other expression that is not "just" a table column)
* Handle constants
* Handle unnamed fields, like the result of the sql function or the constant.

Reply:
@weiznich, that's exactly what I'm trying to convey to you. Why do you report this example only here in the comments, but you cannot understand that it should be in the documentation? The very fact of the presence of this PR, [as](https://github.com/diesel-rs/diesel/issues/2475) [well](https://github.com/diesel-rs/diesel/issues/2169) [as](https://github.com/diesel-rs/diesel/issues/2011) [many](https://github.com/diesel-rs/diesel/issues/1926) [and](https://github.com/diesel-rs/diesel/issues/1518) [many](https://github.com/diesel-rs/diesel/issues/1472) [other](https://github.com/diesel-rs/diesel/issues/573) [issues](https://github.com/diesel-rs/diesel/issues/421), confirms that the behavior of `Queryable` is incomprehensible not only to me.

Besides, in the [getting started guide](http://diesel.rs/guides/getting-started/) (which will be one of the first places to get information of the idiomatic diesel code) we see:

> Using `#[derive(Queryable)]` assumes that the order of fields on the `Post` struct matches the columns in the `posts` table, so make sure to define them in the order seen in the `schema.rs` file.

and in the API docs:
1.4.x: https://github.com/diesel-rs/diesel/blob/bec389a1ea329c161fc7a8fb8df8bea29a3432bf/diesel/src/deserialize.rs#L27-L30
master: https://github.com/diesel-rs/diesel/blob/8f149f9cdc085974283dc6e997390ed51bf7ec19/diesel_derives/src/lib.rs#L397-L400

If `Queryable` is not connected to the `table!` macro at all how order can be significant? And order of what? Can you specify it for your own example? But if `Queryable` is connected to `table!`, why you should keep order in sync manually? Compiler have enough information to do that for you.

Reply:
Ok, again: I belive both of you @Mingun and @pickfire are missing a few really important point here:

* First of all, diesel never claimed to work like other ORM's. If people believe that and try to apply their knowledge from say active record here, they will fail. (In the end that's similar to rust itself, if you try to squeeze your object orientated knowledge in there you will run into the borrow checker really fast). So please stop to transfer some knowledge that you have about ORM models to `Queryable`. If it where the same we would just have named it `Model` instead.
* Our documentation is a place to lookup things for all users, not only for beginners. Therefore we cannot simplify things here, as this would block more advanced usages of our API. All formulations give so far only target a single "problem" you've hit while starting diesel. It's great that you try to improve the documentation there, but please accept that I want to keep things written in our documentation true for all cases, not only for a single edge case.
* As pointed out multiple times above: Both of you seem to belief that `Queryable` is something magical. It is not, its doing something really simple. For me it seems like you attribute nearly all behaviour coming from a query dsl method call chain to `.load()` or `Queryable`, while in practice each of the parts have a separate influence. In the end that's the reason why I'm so hasty to add all this to the documentation of `Queryable`/`#[derive(Queryable)]`. It's just not the right place for this in my opinion. 
(I've already pointed this out [here](https://github.com/diesel-rs/diesel/pull/2631#discussion_r564534126))

Now I could say, technically it's clearly documented how `Queryable` works, as this is part of the trait bounds of our [public API](https://docs.diesel.rs/1.4.x/diesel/query_dsl/methods/trait.LoadQuery.html#implementors) (that's just on click away from `RunQueryDsl::load()`). Technically you just need to read those bounds and everything should be clear. I totally understand that this can be challenging and adding some more explicit documentation won't hurt, therefore here a short explanation on how I understand that query able works:

Let `Q` be some select query, that has a select clause `(S1, S2, …, Sn)`. Each select clause is required to be a valid `Expression`, therefore we can determine the SQL type of that select clause as `(ST1, ST2, …, Sn)` where `S1: Expression<SqlType = ST1>` and so on. For `load` it is now required to have a `Queryable` impl of the form `Queryable<(ST1, ST2, …, Sn), DB>` for your type where `DB` is just the used backend.  Now let's assume that you have the you want to derive `Queryable` for the following type:

```rust
struct MyType {
    first_field: T1,
    second_field: T2,
    …
    last_field: Tn,
}
```

All `#[derive(Queryable)]` does now is providing the following impl:
(using the `Queryable` definition from the master branch here, but the idea is the same for the 1.4.x releases.
```rust
impl<ST1, ST2, …, STn, DB> Queryable<(ST1, ST2, …, STn), DB> for MyType {
    type Row = (T1, T2, …, Tn);
    
    fn row((first_field, second_field, …, last_field): Self::Row) -> Result<Self> {
        Ok(Self { first_field, second_field, …, last_field } )
    }
}
```

As you easily can see does this impl nothing magical. It's just a stupid simple field mapping, that does not even know about `table!` or whatever. In my opinion most of the problems you've listed above are coming from people that do not understand how the select clause is composed, and not from how `#[derive(Queryable)]` work. That's just where the misunderstand surfaces, not where the actual issue lies. It is certainly possible to fix the underlying problem with documentation, but as pointed out above already: The API documentation on the relevant functions is already good in my opinion, what is missing is a overall guide that shows how this functions interact with each other. As already pointed out multiple times: I would be happy if someone writes such a guide, but as this is really a lot of work and I'm doing most of the work on diesel in my free time I'm not willing to spend that time on something that I cannot use for my own projects. 

So this is likely my last comment to this PR, at least as long as the changes do not reflect what I've written above.

Reply:
Oh, I think I get it now after your explanation but I think things are still missing in the documentation, at least not in the `Queryable` trait part since that is pretty clear.

https://docs.diesel.rs/1.4.x/diesel/query_dsl/trait.RunQueryDsl.html#returning-a-struct

Because of that example didn't mention that the struct needs to have all the fields in the table so that's why I am confused.

> Types which implement Queryable represent the result of a SQL query. This does not necessarily mean they represent a single database table.

Combine that with `Queryable` trait doc stuff become more confusing. From my point of view, the example above to load struct thinks of the struct as a "model" rather than something "queryable". A "model" needs to have all the fields but "queryable" does not need to have all the fields.

Maybe within the code we can mention that that only works if the `Queryable` struct have all the fields in the table in the same order. Do you think that is good to remind others a second time?

Reply:
> Because of that example didn't mention that the struct needs to have all the fields in the table so that's why I am confused.

Because again: For the general case that's just not true. You cannot write there that the return type will need to have all fields of an table, because that opens a lot of questions about which table exactly, how this works with select clauses and so on.

>  From my point of view, the example above to load struct thinks of the struct as a "model" rather than something "queryable". A "model" needs to have all the fields but "queryable" does not need to have all the fields.

Can you point at some of our documentation mentioning the term "model". I'm quite sure that we explicitly avoid this term for this reason. For me that seems more like you are trying to apply your knowledge from different ORM's to diesel.

> Maybe within the code we can mention that that only works if the `Queryable` struct have all the fields in the table in the same order.
 
Within which code? Also I see using the formulation " all the fields in the table in the same order" critically, as I already know that someone will then try to use literally `users::name` as struct field type or so.

Reply:
```rust
#[derive(Queryable, PartialEq, Debug)]
struct User {
    id: i32,
    name: String,
}

let data = users
    .load::<User>(&connection)?;
let expected_data = vec![
    User { id: 1, name: String::from("Sean"), },
    User { id: 2, name: String::from("Tess"), },
];
assert_eq!(expected_data, data);
```
The example only shows that, people may not notice that `User` includes every item from `users` table, to be used this way. I mean maybe the example is lacking some context. It did not mention the word "model" but it feels like a model to me, maybe just my view, so I did the same thing (plus less fields) and it didn't work.

Reply:
In my understanding this example just points out that instead of receiving a tuple from the database (as shown in the example above) you could also map that to a struct containing the same fields in the same order. Having written that I can at least understand why you got that impression, so maybe that could be pointed out there explicitly?

Reply:
Yes, if we could have a comment saying that it is a **`Queryable` struct containing the same fields in the same order** that would be very good. At least others know that if it is just a normal struct or does not have all the fields or not in the same order would not work. It could give readers a clear context and the conditions required to do so.

Another suggestion I have is we could show an example on using `(User, Post)` in either `load` or `inner_join` docs since struct load is shown in `load` (without `explict` select), and we could mention the same thing in the above paragraph. I could send a pull request for this if you think it is a good idea (like the example above but with join).

Reply:
> Yes, if we could have a comment saying that it is a Queryable struct containing the same fields in the same order that would be very good. At least others know that if it is just a normal struct or does not have all the fields or not in the same order would not work. It could give readers a clear context and the conditions required to do so.

Adding such a statement to the documentation of `load` seems fine as long as "same fields" refers explicitly to tuple variant above.

> Another suggestion I have is we could show an example on using (User, Post) in either load or inner_join docs since struct load is shown in load (without explict select), and we could mention the same thing in the above paragraph. I could send a pull request for this if you think it is a good idea (like the example above but with join).

I believe that should be added to the examples section in the documentation of `inner_join`.

Reply:
I have updates the examples for this.